plugins/reindex/src/test/java/org/elasticsearch/plugin/reindex/UpdateByQueryWhileModifyingTest.java

```diff
@@ -0,0 +1,103 @@
+ */
+public class UpdateByQueryWhileModifyingTest extends UpdateByQueryTestCase {
+    private static final int MAX_MUTATIONS = 50;
+    private static final int MAX_ATTEMPTS = 10;
+
+    public void testUpdateWhileReindexing() throws Exception {
+        AtomicReference<String> value = new AtomicReference<>(randomSimpleString(random()));
+        indexRandom(true, client().prepareIndex("test", "test", "test").setSource("test", value.get()));
+
+        AtomicReference<Throwable> failure = new AtomicReference<>();
+        AtomicBoolean keepUpdating = new AtomicBoolean(true);
+        Thread updater = new Thread(new Runnable() {
+            @Override
+            public void run() {
+                while (keepUpdating.get()) {
+                    try {
+                        assertThat(request().source("test").refresh(true).get(),
+                                responseMatcher().updated(either(equalTo(0L)).or(equalTo(1L)))
+                                        .versionConflicts(either(equalTo(0L)).or(equalTo(1L))));
+                    } catch (Throwable t) {
+                        failure.set(t);
+                    }
+                }
+            }
+        });
+        updater.start();
+
+        try {
+            for (int i = 0; i < MAX_MUTATIONS; i++) {
+                GetResponse get = client().prepareGet("test", "test", "test").get();
+                assertEquals(value.get(), get.getSource().get("test"));
+                value.set(randomSimpleString(random()));
+                IndexRequestBuilder index = client().prepareIndex("test", "test", "test").setSource("test", value.get())
+                        .setRefresh(true);
+                /*
+                 * Reindex increments the version number so concurrent indexes
+                 * might get version conflict exceptions so we just blindly
+                 * retry.
+                 */
+                int attempts = 0;
+                while (true) {
+                    attempts++;
+                    try {
+                        index.setVersion(get.getVersion()).get();
+                        break;
+                    } catch (VersionConflictEngineException e) {
+                        if (attempts >= MAX_ATTEMPTS) {
+                            throw new RuntimeException("Failed to index after [" + MAX_ATTEMPTS + "] attempts. Too many version conflicts!");
+                        }
+                        logger.info(
+                                "Caught expected version conflict trying to perform mutation number {} with version {}. Retrying.",
+                                i, get.getVersion());
+                        get = client().prepareGet("test", "test", "test").get();
+                        continue;
+                    }
+                }
+            }
+        } finally {
+            keepUpdating.set(false);
+            updater.join(TimeUnit.SECONDS.toMillis(10));
+            if (failure.get() != null) {
```

Comment:
You're right. That's not good. But you could still throw an `AssertionError` and pass the cause.